core/src/main/java/org/apache/iceberg/BaseFile.java

```diff
@@ -360,7 +360,7 @@ public ByteBuffer keyMetadata() {
     if (list != null) {
       List<E> copy = Lists.newArrayListWithExpectedSize(list.size());
       copy.addAll(list);
-      return Collections.unmodifiableList(copy);
```

Comment:
Revert this?

Reply:
The `AvroSerializer` don't work here because we need the `schema` to serialize and deserialize the `DataFile`,  while in our current design we're hiding the schema details inside  `GenericDataFile` and `BaseDataFile`.  The current kyro serializer don't have problem actually, so I'd prefer to keep the current version.

Reply:
Can we register something with Kryo to work with the unmodifiable list?

Reply:
+1 We can register a custom serializer to Flink.

Reply:
@rdblue @JingsongLi   I followed the [suggestion](https://stackoverflow.com/questions/32453030/using-an-collectionsunmodifiablecollection-with-apache-flink) from flink community and prepared a patch for the  serializer issue [here](https://github.com/apache/iceberg/pull/1185/commits/b5b518b4ddcb36d9219c7e53cf40704f3a2e0e4e),  How do you think about this patch ?

Reply:
For me, it sounds not a good solution for our iceberg here because: 
1.  it introduced a java dependency from an individual developer (https://github.com/magro/kryo-serializers).
2.  It use a hack way (loading an invisible class) to register the UnmodifiableCollectionsSerializer.

Reply:
Sorry if it's inappropriate for me to but in here, as I'm not a committer or part of the project.

But I maintain and help developers at my company with dozens of flink programs. I can attest that we have had issues with using `de.javakaffee:kryo-serializers` in our flink programs, specifically with UnmodifiableListSerializer. As one example, flink allows users to override the class loading to either be parent first or child first. This, combined with the loading of an invisible class via `Class.forName` to register the UnmodifiableCollectionsSerializer has lead to developers encountering class not found issues depending on how they compile their programs.

Additionally, in the newer versions of flink (I belive >1.9, which current is 1.11 recently released), certain internal classes use their own class loading order, ignoring the configured class loading order (parent vs child) which the user configured. This has lead to some unexpected issues at run time, namely class not found errors as the class is not on the correct class path for flink's kryo dependency to find it.

In general, kryo is typically attempted to be avoided in flink-landia in favor of custom serializers which use the Flink serializer / deserializer interface.

So I'd +1 with @JingsongLi to registering a custom serializer a custom flink serializer to be used in the flink-iceberg subproject.

Reply:
Additionally, I've not contributed meaningfully to the code base (I fixed some white space issues), but I have been following along for quite some time and I'm very interested in attempting to using Flink to write Iceberg files at my work. If there's any way I could be of assistance with this, please let me know. Possibly I could help fill in the gaps of Flink knowledge.

Please let me know where a better place to discuss new contributor opportunities would be! But I thought it would be prudent to share my experience with users using `de.javakaffee:kryo-serializers` in their flink programs.

I'd also be open to testing out flink + iceberg integration if there's a need for that.

Reply:
Thanks for your  sharing, @kbendick.  So let's remove the kyro serializer way, and turn to register a custom serializer.

Reply:
No problem @openinx. There's essentially an issue with kryo's `JaveSerializer` when used with flink. See here: https://issues.apache.org/jira/browse/FLINK-6025 and here for more info: https://ci.apache.org/projects/flink/flink-docs-stable/dev/custom_serializers.html#issue-with-using-kryos-javaserializer

If there's anyway I could be of assistance with this issue or any other work for someone hopeful to start contributing to the project in a meaningful way, I'd be very happy to. I'm personally evaluating iceberg for usage between flink and spark as there are many flink users that I help oversee and we also have need to integrate that work with spark. 

I'd be happy to attempt a PR for a custom serializer, or would be happy to work on anything that anybody feels would be good for a hopeful new committer / contributor. Including tests, automation set up, flink support, etc.

Sorry if there's a better place to find this information; I am just getting my hands dirty in contributing to an apache OSS project (though I have forked some and then made changes and used those forks in production). I've also browsed the issues etc, but I'm not quite sure how to go about being allowed to get assigned one. My apologies if this is covered elsewhere; I'm used to being able to browse the JIRA for apache projects (though I much prefer github).

cc @JingsongLi for potential guidance if you have a minute :)